File: LeetCode/Medium/0002_Add_Two_Numbers/Solution.java
// 2021-06-01
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return generateNode(l1, l2, 0);
    }
    
    private ListNode generateNode(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null && carry == 0) return null;
        if (l1 == null && l2 == null && carry > 0) return new ListNode(carry, null);
        
        if (l1 == null) {
            int sum = l2.val + carry;
            int value = sum % 10;
            carry = sum / 10;
            return new ListNode(value, generateNode(l1, l2.next, carry));
        }
        
        if (l2 == null) {
            int sum = l1.val + carry;
            int value = sum % 10;
            carry = sum / 10;
            return new ListNode(value, generateNode(l1.next, l2, carry));
        }
        
        int sum = l1.val + l2.val + carry;
        int value = sum % 10;
        carry = sum / 10;    
        return new ListNode(value, generateNode(l1.next, l2.next, carry));
        
    }
}

File: LeetCode/Medium/0002_Add_Two_Numbers/Solution.py
from typing import *
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

## 2022-11-03
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = num_sum = ListNode(-1)
        carry = 0
        while l1 and l2:
            next_sum = l1.val + l2.val + carry
            carry, val= next_sum//10, next_sum%10
            num_sum.next = ListNode(val)
            l1, l2, num_sum = l1.next, l2.next, num_sum.next
        
        remain = l1 if l1 else l2
        while remain:
            next_sum = remain.val + carry
            carry, val= next_sum//10, next_sum%10
            num_sum.next = ListNode(val)
            remain, num_sum = remain.next, num_sum.next
            
        if carry:
            num_sum.next = ListNode(1)

        return head.next
